Upgrade to eslint4

leo-shopify opened the pull request:
- Fixes the difference in `lib/config/core.js` that was preventing v4 from running.
- Sets the minimum supported node version to `6.11.1` (LTS).
- Sets `indent-legacy` and disables new `indent` rule for backwards compat.
- Updates dependencies.
- Adds pre/postversion scripts to make publishing a bit more convenient.
- The next release should be a **major** version update.

## 🎩 with a pre-release version
1. Open a local repo
2. Make sure it lints without errors or warnings using v16.0.1
3. `yarn add eslint https://github.com/Shopify/eslint-plugin-shopify/archive/v17.0.0-1.tar.gz`

## Merge checklist
- [x] Fix `can't resolve reference #/definitions/order from id #`.
    - ~~Waiting for upstream fix bryanrsmith/eslint-plugin-sort-class-members#41.~~
- [x] Top hat a few existing repositories using the pre-release version.
    - [x] [sell-on-amazon](https://github.com/Shopify/sell-on-amazon)
    - [x] [identity](https://github.com/Shopify/identity)

## Release checklist
Whomever has npm publishing rights, before publishing should:
- [ ] Verify the release date in the [CHANGELOG](https://github.com/Shopify/eslint-plugin-shopify/blob/upgrade-to-eslint4/CHANGELOG.md#1700---2017-08-09)
- [ ] Run `yarn version --new-version 17.0.0`.

Commit:
update eslint version

- update configuration to new structure
- update docs

Commit:
fix eslint warnings

Commit:
normalize deps version specification; upgrade deps; add missing peer deps

`react` and `react-dom` were missing peer dependencies of `redbox-react`

Commit:
add `preversion` and `postversion` scripts

Test and lint before updating the package version and push tags afterwards.

Commit:
use legacy indent rules

See http://eslint.org/docs/user-guide/migrating-to-4.0.0#indent-rewrite

Commit:
set minimum node version to current LTS version

The node version in circle was too old and failing:

`yarn run check`

> yarn run v0.27.5
> $ yarn run lint && yarn test
> can't resolve reference #/definitions/order from id #
> can't resolve reference #/definitions/order from id #
> /home/ubuntu/eslint-plugin-shopify/node_modules/babel-preset-shopify/web.js:4
> module.exports = function shopifyWebPreset(context, options = {}) {

See: https://circleci.com/gh/Shopify/eslint-plugin-shopify/114

GoodForOneFare commented:
I've narrowed down the `can't resolve reference` messages to `node_modules/eslint-plugin-shopify/lib/config/rules/sort-class-members.js`.

leo-shopify commented:
The error is from json schema. I sent a PR to the maintainer: bryanrsmith/eslint-plugin-sort-class-members#41

Commit:
remove white-space

leo-shopify commented:
The other error I was getting: `no-multi-assign` being reported, was because the repos I was testing were on v15.x and this rule was added on v.16 26d6b79.
I removed it from the description.

ismail-syed commented on package.json:
```diff
@@ -37,33 +42,35 @@
+    "eslint-plugin-chai-expect": "^1.1.1",
+    "eslint-plugin-flowtype": "^2.35.0",
+    "eslint-plugin-import": "^2.7.0",
+    "eslint-plugin-jsx-a11y": "^6.0.2",
```
`eslint-plugin-jsx-a11y`'s [`5.1.0`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/CHANGELOG.md#510--2017-06-26) release added a few new rules. I think we should include and disabled them at the very least.

leo-shopify replied:
My reason for upgrading was to normalize the version specifiers, which I thought was accidental but your other comment suggests it might have been on purpose.

I'd prefer to track any rule updates that are not necessary to upgrade to v4 separately.

Maybe it's better to revert the dependencies?

ismail-syed replied:
> I'd prefer to track any rule updates that are not necessary to upgrade to v4 separately.

I agree. 

> Maybe it's better to revert the dependencies?

I would suggest reverting deps that aren't relevant to the v4 upgrade.

ismail-syed commented on package.json:
```diff
@@ -37,33 +42,35 @@
-    "eslint-plugin-lodash": "^2.4.2",
-    "eslint-plugin-mocha": "^4.9.0",
-    "eslint-plugin-node": "^4.2.2",
+    "eslint-plugin-ava": "^4.2.1",
```
Should we allow patch updates on all our plugins? Previously some weren't and some were. Do we remember the reasoning behind that?

ismail-syed commented on package.json:
```diff
@@ -10,14 +10,19 @@
-    "test:cover": "babel-node node_modules/.bin/isparta cover --report text --report html _mocha -- --reporter spec --recursive tests/"
+    "test:cover": "babel-node node_modules/.bin/isparta cover --report text --report html _mocha -- --reporter spec --recursive tests/",
+    "preversion": "yarn run check",
+    "postversion": "git push && git push --tags"
```
Do the `preversion` and `postversion` scripts run when we do something like: 
```
yarn version --new-version patch
```

Or when we do 
``` 
yarn run version
```

ismail-syed replied:
Also, I prefer 
```
git push --follow-tags
```

I get the most consistent shipit behavior with that.

leo-shopify replied:
Yes, they both run when a new version is set, but not when queried. i.e. `npm version`.

From https://docs.npmjs.com/misc/scripts#description:
- preversion: Run BEFORE bumping the package version.
- version: Run AFTER bumping the package version, but BEFORE commit.
- postversion: Run AFTER bumping the package version, and AFTER commit.

ismail-syed replied:
Ahh okay, my concern is that code/tags accidentally gets pushed when a dev does some version tweaking locally. 

Eg. Accidentally doing a minor bump instead of a patch.  


Also, I don't think the preversion is necessary. We deploy through shipit, which expects CI to pass before deploying. Our CI already runs `yarn run check`.

ismail-syed reviewed (commented):
On master, I'm able to lint this repo with itself. I can't seem to do that on this branch?

```
yarn run lint
```

results in:
```
╰─$ yarn run lint
yarn run v0.21.3
$ eslint . --max-warnings 0
Error: [object Object]:
	ESLint configuration is invalid:
	- Unexpected top-level property "ecmaFeatures".

Referenced from: /Users/ismailsyed/src/github.com/Shopify/eslint-plugin-shopify/node_modules/eslint-plugin-shopify/index.js
Referenced from: /Users/ismailsyed/src/github.com/Shopify/eslint-plugin-shopify/package.json
```

Any ideas?

Commit:
Use `--follow-tags` instead of `--tags` on `postversion`

See https://github.com/Shopify/eslint-plugin-shopify/pull/34#discussion_r131382224

leo-shopify commented:
@ismail-syed fwiw running the linter on itself works fine for me:
> yarn run lint
> yarn run v0.21.3
> $ eslint . --max-warnings 0 
> ✨  Done in 2.98s.

Try `rm -r node_modules yarn.lock && yarn install`

ismail-syed commented:
>  Try rm -r node_modules yarn.lock && yarn install

Still getting the same issue 😞

Commit:
remove updates that added new rules

Commit:
release pre-major

Commit:
v17.0.0-1

leo-shopify commented:
- The dependency that was causing the error has been fixed.
- Reverted deps version to only minor updates
- Bumped the pre-release version for top-hatting.

beefchimi commented:
Currently using this latest version in a project and have yet to encounter any issues 👍

ismail-syed commented on package.json:
```diff
@@ -10,14 +10,19 @@
-    "test:cover": "babel-node node_modules/.bin/isparta cover --report text --report html _mocha -- --reporter spec --recursive tests/"
+    "test:cover": "babel-node node_modules/.bin/isparta cover --report text --report html _mocha -- --reporter spec --recursive tests/",
+    "preversion": "yarn run check",
+    "postversion": "git push && git push --follow-tags"
```
Sorry for not being clear enough before. 
I meant replacing `git push && git push --follow-tags` with `git push --follow-tags`. 

`git push --follow-tags` will push your commits along with new tags.

ismail-syed commented on CHANGELOG.md:
```diff
@@ -1,6 +1,17 @@
 # Changelog
 
 ## [Unreleased]
+### Changed
```
Are we planning to ship a release for `v17.0.0` once this is merged?
If so, let's add a header for `v17.0.0` along with a date in this PR as well.

leo-shopify replied:
My assumption was that the person with npm publishing credentials would decide.
I'll add it anyway.

ismail-syed replied:
Thanks 👍

ismail-syed commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "eslint-plugin-shopify",
-  "version": "16.0.1",
+  "version": "17.0.0-1",
```
Will this be reverted back to `16.0.1` before merging?

I believe we leave this untouched. When we 🚢 a release, `yarn version --new-version <major|minor|patch>` will bump it up for us.

leo-shopify replied:
The trailing `-number` makes it a pre-release, which has lower precedence than a `17.0.0`.
Once we decide to release it, whomever is publishing should run `yarn version --new-version 17.0.0`.

Commit:
Correct postversion script

See https://github.com/Shopify/eslint-plugin-shopify/pull/34#pullrequestreview-55007853

Commit:
set the changelog version and estimated release date

See https://github.com/Shopify/eslint-plugin-shopify/pull/34#pullrequestreview-55008800

ismail-syed reviewed (approved):
LGTM, thanks again!
💯

GoodForOneFare commented:
Taking another look now.

GoodForOneFare commented:
Tweaking `shopify/shopify` to use `eslint@4.3.0` + this branch, I see these messages when linting `shopify/shopify`:

```
can't resolve reference #/definitions/basicConfig from id #
can't resolve reference #/definitions/basicConfigOrBoolean from id #
can't resolve reference #/definitions/basicConfigOrBoolean from id #
```

Looks like those are coming from `node_modules/eslint-plugin-react/lib/rules/jsx-curly-spacing.js` I've verified that `eslint-plugin-shopify` and `shopify/shopify` are both using `eslint-plugin-react@7.1.0`.

GoodForOneFare commented on package.json:
```diff
@@ -37,33 +42,35 @@
+    "eslint-plugin-chai-expect": "^1.1.1",
     "eslint-plugin-flowtype": "^2.32.1",
-    "eslint-plugin-import": "2.2.x",
+    "eslint-plugin-import": "^2.2.0",
```
Not a blocker for this PR, but `shopify/shopify` is going to fail `yarn check` until a new `eslint-plugin-import` release ([it will loosen up the expected dependencies](https://github.com/benmosher/eslint-plugin-import/commit/72d13858e044569d6f194922e507578410601f28)).

Seeing:
```
error "\u001b[2meslint-plugin-shopify#\u001b[22meslint-plugin-import#eslint@2.x - 3.x" doesn't satisfy found match of "eslint@4.3.0"
```

GoodForOneFare commented on circle.yml:
```diff
@@ -1,6 +1,8 @@
   environment:
     PATH: "${PATH}:${HOME}/${CIRCLE_PROJECT_REPONAME}/node_modules/.bin"
+  node:
+    version: 6.11.1 # current is 8.2.1
```
This comment feel superfluous, and will become stale quickly.  We try to standardise on the LTS version, which is `6.x` until October.

GoodForOneFare reviewed (approved):
Hold off on 🚢  until `eslint-plugin-react`'s messages are gone.

Aside from the node version comment , everything LGTM 👍

Commit:
remove leftover comment

see https://github.com/Shopify/eslint-plugin-shopify/pull/34#discussion_r132015547

leo-shopify commented:
@GoodForOneFare Ping me when the messages are gone.

ismail-syed commented:
Any updates on this?

GoodForOneFare commented:
Someone has to clean up the messages that I listed in my previous comment.  It likely just requires [1 line PRs like Leo's](https://github.com/bryanrsmith/eslint-plugin-sort-class-members/pull/41/files) to affected repos.

GoodForOneFare committed:
Update min react & sort-class-members libs

This gets rid of errors like, `The missing `id` property generates `can't resolve reference #/definitions/order from id #`.

GoodForOneFare commented:
@leo-shopify I pushed a couple of version bumps to get rid of those messages.  After 🎩 in admin, this looks good to 🚢!

Thank you so much for pushing this through ❤️

leo-shopify commented:
My pleasure!
Next steps to publish:
- [ ] Verify the release date in the [CHANGELOG](https://github.com/Shopify/eslint-plugin-shopify/blob/upgrade-to-eslint4/CHANGELOG.md#1700---2017-08-09)
- [ ] Run `yarn version --new-version 17.0.0`.

ismail-syed commented:
@leo-shopify 
Yup, and then deploy through shipit.

leo-shopify commented:
@ismail-syed Just so it doesn't get stuck in limbo:
I'm assuming you or whomever has the npm credential will do these last steps.
Of course, ping me if anything else is needed from me but otherwise my work here should be done.

GoodForOneFare commented:
@leo-shopify everyone has credentials via https://shipit.shopify.io/shopify/eslint-plugin-shopify/production :)  Just click the deploy button when you have a new version number + tag.

leo-shopify commented:
Nice! I was not aware of all this voodoo. Thanks for the heads-up!
https://www.npmjs.com/package/eslint-plugin-shopify
